File: c/demo/quickstart.cpp
#include <revo3/revo3.hpp>

#include <chrono>
#include <cstdio>
#include <cstring>
#include <stdexcept>
#include <thread>
#include <vector>

namespace {

enum class MoveScope { None, Hand, Joint, Finger, Thumb };
constexpr std::uint8_t kSafetyRecoveryRequired = 1;
constexpr std::uint8_t kSafetyFaulted = 2;

const char *model_name(Revo3Model model) {
  switch (model) {
  case REVO3_MODEL_ULTRA:
    return "Revo3 Ultra";
  case REVO3_MODEL_ULTRA_TOUCH:
    return "Revo3 Ultra Touch";
  case REVO3_MODEL_ULTRA_VISION_TOUCH:
    return "Revo3 Ultra VisionTouch";
  case REVO3_MODEL_PRO:
    return "Revo3 Pro";
  case REVO3_MODEL_PRO_TOUCH:
    return "Revo3 Pro Touch";
  case REVO3_MODEL_BASIC:
    return "Revo3 Basic";
  case REVO3_MODEL_BASIC_TOUCH:
    return "Revo3 Basic Touch";
  default:
    return "Revo3";
  }
}

bool health_has_fault(const revo3::HealthSnapshot &health) {
  return health.safety_state == kSafetyRecoveryRequired ||
         health.safety_state == kSafetyFaulted || health.system_state != 0 ||
         health.system_error_code != 0 || health.faulted_motor_count != 0;
}

const char *operation_state_name(revo3::OperationState state) {
  switch (state) {
  case revo3::OperationState::Pending:
    return "Pending";
  case revo3::OperationState::Running:
    return "Running";
  case revo3::OperationState::Succeeded:
    return "Succeeded";
  case revo3::OperationState::Cancelled:
    return "Cancelled";
  case revo3::OperationState::Preempted:
    return "Preempted";
  case revo3::OperationState::Failed:
    return "Failed";
  case revo3::OperationState::Indeterminate:
    return "Indeterminate";
  }
  return "Unknown";
}

}  // namespace

int main(int argc, char **argv) {
  using namespace std::chrono_literals;

  try {
    revo3::DiscoveryOptions discovery;
    MoveScope move_scope = MoveScope::None;
    bool allow_unhealthy = false;
    for (int index = 1; index < argc; ++index) {
      if (std::strcmp(argv[index], "--help") == 0 || std::strcmp(argv[index], "-h") == 0) {
        std::printf(
            "Usage: %s [--port <name>] [--move | --move-joint | --move-finger | --move-thumb] "
            "[--strict-health | --allow-unhealthy]\n\n"
            "Options:\n"
            "  --port <name>      Optional serial or CAN port (e.g. /dev/ttyUSB0)\n"
            "  --move             Move the whole hand (4-finger flex & return)\n"
            "  --move-joint       Move a single joint (joint 0)\n"
            "  --move-finger      Move a single finger (index finger)\n"
            "  --move-thumb       Move the thumb\n"
            "  --strict-health    Refuse motion when Health/State reports fault (default)\n"
            "  --allow-unhealthy  Allow motion despite Health/State warnings\n",
            argv[0]);
        return 0;
      } else if (std::strcmp(argv[index], "--port") == 0 && index + 1 < argc) {
        discovery.port = argv[++index];
      } else if (std::strcmp(argv[index], "--move") == 0) {
        move_scope = MoveScope::Hand;
      } else if (std::strcmp(argv[index], "--move-joint") == 0) {
        move_scope = MoveScope::Joint;
      } else if (std::strcmp(argv[index], "--move-finger") == 0) {
        move_scope = MoveScope::Finger;
      } else if (std::strcmp(argv[index], "--move-thumb") == 0) {
        move_scope = MoveScope::Thumb;
      } else if (std::strcmp(argv[index], "--strict-health") == 0) {
        allow_unhealthy = false;
      } else if (std::strcmp(argv[index], "--allow-unhealthy") == 0) {
        allow_unhealthy = true;
      } else {
        discovery.port = argv[index];
      }
    }

    revo3::Manager manager;
    auto hand = manager.connect_auto(discovery);

    const auto device_info = hand.device_info();
    const auto firmware_info = hand.firmware_info();
    const auto layout = hand.joint_layout();
    if (!layout) {
      throw std::runtime_error("Joint layout is unavailable");
    }

    std::printf("Device:   %s (%s)\n",
                device_info.serial_number.empty() ? "unknown" : device_info.serial_number.c_str(),
                device_info.hand_side == revo3::HandSide::Right ? "Right" : "Left");
    std::printf("Slave ID: %u\n", hand.slave_id());
    std::printf("Model: %s | Hardware revision: %s | Firmware: %s\n",
                model_name(static_cast<Revo3Model>(device_info.model)),
                device_info.hardware_revision.empty() ? "unknown" : device_info.hardware_revision.c_str(),
                firmware_info.controller_firmware_version.empty() ? "unknown" : firmware_info.controller_firmware_version.c_str());
    std::printf("Layout:   %s (%u DOF)\n\n", layout->layout_id.c_str(),
                layout->joint_count);

    const auto state = hand.state().snapshot();
    std::printf("State received: J0=%.2f degree, J0 velocity=%.2f rpm, "
                "J0 current=%.2f mA\n",
                state.motors.positions_deg[0], state.motors.velocities_rpm[0],
                state.motors.currents_ma[0]);
    const auto health = hand.health().snapshot();
    std::printf("Health: system_state=%u system_error=%u faulted_motor_count=%u\n",
                health.system_state, health.system_error_code,
                health.faulted_motor_count);

    if (move_scope != MoveScope::None) {
      const bool health_fault = health_has_fault(health);
      if (!allow_unhealthy && health_fault) {
        throw std::runtime_error(
            "Refusing to move because strict Health preflight found a fault");
      }
      if (health_fault) {
        std::printf("Warning: Health preflight found safety=%u "
                    "system_state=%u system_error=%u faulted_motor_count=%u; "
                    "continuing because --allow-unhealthy is set.\n",
                    health.safety_state, health.system_state,
                    health.system_error_code, health.faulted_motor_count);
      }
      if (layout->joint_count != 21) {
        throw std::runtime_error("Motion currently requires a 21-joint layout");
      }

      if (move_scope == MoveScope::Hand) {
        std::printf("Phase 1: Flexing fingers to 45.0 deg...\n");
        std::vector<float> flex_targets(
            state.motors.positions_deg,
            state.motors.positions_deg + layout->joint_count);
        for (int joint : {1, 2, 5, 6, 9, 10, 13, 14}) {
          flex_targets[static_cast<std::size_t>(joint)] = 45.0F;
        }
        auto motion1 = hand.motion().move_to(flex_targets, 1500ms);
        const auto res1 = motion1.wait(5s);
        std::printf("Motion 1 (Flex): %s (%d)\n", operation_state_name(res1),
                    static_cast<int>(res1));
        if (res1 != revo3::OperationState::Succeeded) {
          return 1;
        }

        std::this_thread::sleep_for(500ms);

        std::printf("Phase 2: Returning fingers to their initial positions...\n");
        std::vector<float> extend_targets(
            state.motors.positions_deg,
            state.motors.positions_deg + layout->joint_count);
        auto motion2 = hand.motion().move_to(extend_targets, 1500ms);
        const auto res2 = motion2.wait(5s);
        std::printf("Motion 2 (Return): %s (%d)\n",
                    operation_state_name(res2), static_cast<int>(res2));
        if (res2 != revo3::OperationState::Succeeded) {
          return 1;
        }
      } else {
        revo3::OperationHandle motion;
        if (move_scope == MoveScope::Joint) {
          motion = hand.motion().move_joint(0, 30.0F, 1500ms);
        } else if (move_scope == MoveScope::Finger) {
          motion = hand.motion().flex_finger(1, 30.0F, 1500ms);
        } else if (move_scope == MoveScope::Thumb) {
          std::vector<float> targets(state.motors.positions_deg + 16,
                                     state.motors.positions_deg + 21);
          targets[1] = 30.0F;
          targets[2] = 30.0F;
          targets[4] = 30.0F;
          motion = hand.motion().move_thumb(targets, 1500ms);
        }
        const auto result = motion.wait(5s);
        std::printf("Motion state: %s (%d)\n", operation_state_name(result),
                    static_cast<int>(result));
        if (result != revo3::OperationState::Succeeded) {
          return 1;
        }
      }
    }

    // Explicit close is optional. RAII closes Hand and Manager on scope exit.
    hand.close();
    return 0;
  } catch (const revo3::SdkError &error) {
    std::fprintf(stderr, "Revo3 error: %s (effect=%u, recovery=%u)\n",
                 error.what(), static_cast<unsigned>(error.operation_effect()),
                 static_cast<unsigned>(error.recovery_requirement()));
    return 1;
  } catch (const std::exception &error) {
    std::fprintf(stderr, "Error: %s\n", error.what());
    return 1;
  }
}

File: python/revo3/quickstart.py
"""Connect to one Revo3 hand through the object-oriented Manager API."""

import argparse
import asyncio

from bc_revo3_sdk import main_mod as sdk

REVO3_MOTOR_FAULT_MASK = (
    (1 << 0)
    | (1 << 1)
    | (1 << 2)
    | (1 << 3)
    | (1 << 4)
    | (1 << 8)
)


def enum_name(value) -> str:
    return str(value).rsplit(".", 1)[-1]


def parse_motor_fault_code(code: int) -> str:
    flags = []
    if code & (1 << 0): flags.append("OverCurrent")
    if code & (1 << 1): flags.append("OverVoltage")
    if code & (1 << 2): flags.append("UnderVoltage")
    if code & (1 << 3): flags.append("OverTemp")
    if code & (1 << 4): flags.append("CurrentSpike")
    if code & (1 << 8): flags.append("Stalled")
    return "|".join(flags) if flags else f"0x{code:04X}"


def active_motor_faults(fault_codes) -> dict[int, str]:
    return {
        idx: parse_motor_fault_code(code)
        for idx, code in enumerate(fault_codes)
        if code & REVO3_MOTOR_FAULT_MASK
    }


def motion_preflight_findings(health) -> list[str]:
    findings = []
    active_faults = active_motor_faults(health.motor_fault_codes)
    if active_faults:
        findings.append(f"motor_faults={active_faults}")
    if health.safety_state in (
        sdk.SafetyState.Faulted,
        sdk.SafetyState.RecoveryRequired,
    ):
        findings.append(f"safety={health.safety_state}")
    if health.system_state != 0:
        findings.append(f"system_state={health.system_state}")
    if health.error_code != 0:
        findings.append(f"error_code={health.error_code}")
    if health.faulted_motor_count != 0:
        findings.append(f"faulted_motor_count={health.faulted_motor_count}")
    return findings


def report_motion_preflight(health, strict: bool) -> None:
    findings = motion_preflight_findings(health)
    if not findings:
        return
    if strict:
        raise RuntimeError(
            "Refusing to move because strict Health preflight found: "
            + ", ".join(findings)
            + ". Resolve the device fault and re-run the read-only check first."
        )
    print(
        "Warning: Health/State preflight found "
        + ", ".join(findings)
        + "; continuing because --allow-unhealthy is set."
    )


async def run(args: argparse.Namespace) -> None:
    print(f"SDK version: {sdk.get_sdk_version()}")
    manager = sdk.Manager()
    hand = None
    try:
        hand = await manager.connect_auto(
            port=args.port,
            slave_id=args.slave_id,
        )

        device_info = hand.device_info
        firmware_info = hand.firmware_info
        layout = hand.joint_layout
        if layout is None:
            raise RuntimeError("Joint layout is unavailable; reconnect with an explicit model")

        sn = device_info.serial_number if device_info else "unknown"
        side = (
            device_info.hand_side.name
            if device_info and hasattr(device_info.hand_side, "name")
            else (str(device_info.hand_side).replace("HandSide.", "") if device_info else "unknown")
        )
        hw_ver = device_info.hardware_revision if device_info and device_info.hardware_revision else "unknown"
        fw_ver = firmware_info.controller_firmware_version or "unknown"
        slave_id = getattr(hand, "slave_id", args.slave_id if args.slave_id is not None else 1)

        print(f"Device: {sn} ({side})")
        model = enum_name(device_info.model) if device_info else "unknown"
        print(f"Slave ID: {slave_id}")
        print(f"Model: {model} | Hardware revision: {hw_ver} | Firmware: {fw_ver}")
        print(f"Layout: {layout.layout_id} ({layout.joint_count} DOF)")
        touch_layout = hand.touch.layout
        touch_status = (
            f"{len(touch_layout.modules)} modules"
            if touch_layout is not None
            else "not available"
        )
        print(f"Touch layout: {touch_status}")
        print()

        snapshot = await hand.state.snapshot()
        print(
            f"State timestamp: {snapshot.timestamp.sec}.{snapshot.timestamp.nsec:09d} "
            f"({snapshot.timestamp.clock})"
        )
        print(f"Positions ({len(snapshot.positions_deg)}): {[round(value, 2) for value in snapshot.positions_deg]}")

        health = await hand.health.snapshot()
        active_errors = active_motor_faults(health.motor_fault_codes)
        print(f"Motor faults: {active_errors or 'None'}")
        print(
            f"Health: safety={health.safety_state}, system_state={health.system_state}, "
            f"error_code={health.error_code}, faulted_motor_count={health.faulted_motor_count}"
        )

        if args.move or args.move_joint or args.move_finger or args.move_thumb:
            report_motion_preflight(
                health,
                strict=args.strict_health or not args.allow_unhealthy,
            )
            ultra_models = {
                hand_type
                for hand_type in (
                    getattr(sdk.Revo3Model, "Ultra", None),
                    getattr(sdk.Revo3Model, "UltraTouch", None),
                    getattr(sdk.Revo3Model, "UltraVisionTouch", None),
                )
                if hand_type is not None
            }
            is_ultra = (
                device_info is not None
                and device_info.model in ultra_models
                and layout.joint_count == 21
            )
            if not is_ultra:
                raise RuntimeError("Motion currently requires a 21-DOF Revo3 Ultra")
            if args.speed is not None and (args.move_finger or args.move_thumb):
                raise RuntimeError("--speed is supported by move_to and move_joint only")
            duration = args.duration
            if duration is None and args.speed is None:
                duration = 1.5
            if args.move_joint:
                motion = await hand.motion.move_joint(
                    args.joint_index,
                    args.angle,
                    duration=duration,
                    speed=args.speed,
                    kp=args.kp,
                    kd=args.kd,
                    dt=args.dt,
                )
            elif args.move_finger:
                motion = await hand.motion.flex_finger(
                    args.finger_index,
                    args.angle,
                    duration=duration,
                    kp=args.kp,
                    kd=args.kd,
                    dt=args.dt,
                )
            elif args.move_thumb:
                thumb_targets = list(snapshot.positions_deg[16:21])
                thumb_targets[1] = args.angle
                thumb_targets[2] = args.angle
                thumb_targets[4] = args.angle
                motion = await hand.motion.move_thumb(
                    thumb_targets,
                    duration=duration,
                    kp=args.kp,
                    kd=args.kd,
                    dt=args.dt,
                )
            else:
                flex_angle = args.angle if args.angle != 30.0 else 45.0
                print(f"Phase 1: Flexing fingers to {flex_angle:.1f}°...")
                targets = list(snapshot.positions_deg)  # degrees
                # Flex MCP (1, 5, 9, 13) and PIP (2, 6, 10, 14) joints for 4 fingers (Pinky..Index)
                for joint in (1, 2, 5, 6, 9, 10, 13, 14):
                    targets[joint] = flex_angle
                motion = await hand.motion.move_to(
                    targets,
                    duration=duration,
                    speed=args.speed,
                    kp=args.kp,
                    kd=args.kd,
                    dt=args.dt,
                )
                state = await motion.wait(timeout=args.timeout)
                print(f"Motion 1 (Flex): {state}")
                if state != sdk.OperationState.Succeeded:
                    if motion.error is not None:
                        raise motion.error
                    raise RuntimeError(f"Flexion motion ended in {state}")

                await asyncio.sleep(0.5)

                print("Phase 2: Returning fingers to their initial positions...")
                extend_targets = list(snapshot.positions_deg)
                motion2 = await hand.motion.move_to(
                    extend_targets,
                    duration=duration,
                    speed=args.speed,
                    kp=args.kp,
                    kd=args.kd,
                    dt=args.dt,
                )
                state2 = await motion2.wait(timeout=args.timeout)
                print(f"Motion 2 (Return): {state2}")
                if state2 != sdk.OperationState.Succeeded:
                    if motion2.error is not None:
                        raise motion2.error
                    raise RuntimeError(f"Extension motion ended in {state2}")
                return
            state = await motion.wait(timeout=args.timeout)
            print(f"Motion {motion.id}: {state}")
            if state != sdk.OperationState.Succeeded:
                if motion.error is not None:
                    raise motion.error
                raise RuntimeError(f"Motion ended in {state}")
    finally:
        if hand is not None:
            await hand.close()
        await manager.close()


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Quickstart demo for Revo3 hand.")
    parser.add_argument("positional_port", nargs="?", help="Optional serial or CAN port (e.g. /dev/ttyUSB0)")
    parser.add_argument("--port", help="Serial or CAN port (e.g. /dev/ttyUSB0)")
    parser.add_argument("--slave-id", type=lambda value: int(value, 0))
    motion = parser.add_mutually_exclusive_group()
    motion.add_argument("--move", action="store_true", help="Move the whole hand")
    motion.add_argument("--move-joint", action="store_true")
    motion.add_argument("--move-finger", action="store_true")
    motion.add_argument("--move-thumb", action="store_true")
    parser.add_argument("--joint-index", type=int, default=0)
    parser.add_argument("--finger-index", type=int, choices=range(1, 5), default=1)
    parser.add_argument(
        "--angle", "--deg", "--flexion",
        dest="angle",
        type=float,
        default=30.0,
        help="Target angle in degrees (e.g. 20 for 20 deg flexion)"
    )
    timing = parser.add_mutually_exclusive_group()
    timing.add_argument("--duration", type=float)
    timing.add_argument("--speed", type=float, help="Whole-hand or joint speed in rpm")
    parser.add_argument("--kp", type=float)
    parser.add_argument("--kd", type=float)
    parser.add_argument("--dt", type=float, default=0.01)
    parser.add_argument("--timeout", type=float, default=5.0, help="Motion timeout in seconds")
    safety = parser.add_mutually_exclusive_group()
    safety.add_argument(
        "--strict-health",
        action="store_true",
        help="Refuse motion when State or Health reports any real fault (default)",
    )
    safety.add_argument(
        "--allow-unhealthy",
        action="store_true",
        help="Allow motion despite Health or State faults after independent verification",
    )
    args = parser.parse_args()
    if not args.port and args.positional_port:
        args.port = args.positional_port
    if (args.kp is None) != (args.kd is None):
        parser.error("--kp and --kd must be provided together")
    return args


if __name__ == "__main__":
    try:
        asyncio.run(run(parse_args()))
    except sdk.SdkError as error:
        print(
            "SDK error: "
            f"code={error.code}, "
            f"effect={error.operation_effect}, "
            f"recovery={error.recovery_requirement}, "
            f"retryable={error.retryable}, message={error.message}"
        )
        if error.operation_effect == sdk.OperationEffect.Indeterminate:
            print("Command result is unknown; read state before deciding whether to retry")
        raise SystemExit(1) from error
    except RuntimeError as error:
        print(f"Error: {error}")
        raise SystemExit(1) from error
